src/block/block_compact.c

```diff
@@ -539,9 +531,17 @@ __compact_page_skip(
     else
         ++block->compact_pages_rewritten;
 
-    /* Estimate how much work is left. */
-    if (block->compact_pages_rewritten_expected == 0)
+    /*
+     * We must have reviewed at least some interesting number of pages for any estimates below to be
+     * worthwhile.
+     */
+    if (block->compact_pages_reviewed >= WT_THOUSAND &&
+      block->compact_pages_rewritten_expected == 0) {
         __block_compact_estimate_remaining_work(session, block);
+        /* If no potential work has been found, exit compaction. */
+        if (block->compact_pages_rewritten_expected == 0)
```

Comment:
This check might not be enough to cover more scenarios. With the current reproducer, the writers fill 100% of the free space that was there at the start of compaction. What if they cover 50% of the free space? We may rewrite only a percentage of the blocks, should we have a check `btree_compact_pages_skipped` vs `btree_compact_pages_rewritten`?

Reply:
I think this should cover it? At this point, the estimation function does not have early exit, so exiting based on `compact_pages_rewritten_expected` is meaningful. I don't understand your question about 50%; how would it change the behavior here?

The only question in my mind is whether we could get a call to `__wt_block_compact_progress` after setting `session->compact_state = WT_COMPACT_EXITING` but before actually exiting the compaction. If there is such possibility, it would unnecessarily redo the estimation. 

If there is any doubt, feel free to add another field to `WT_BLOCK` that marks whether the estimation has been done for the file, which would be a more robust way to do this.

Reply:
> I don't understand your question about 50%; how would it change the behavior here?

I have DM'ed you about it. Once it is clearer, I will write a summary here.

> The only question in my mind is whether we could get a call to __wt_block_compact_progress after setting session->compact_state = WT_COMPACT_EXITING but before actually exiting the compaction. If there is such possibility, it would unnecessarily redo the estimation.

That's a very good point, I will verify this with the reproducer.

> If there is any doubt, feel free to add another field to WT_BLOCK that marks whether the estimation has been done for the file, which would be a more robust way to do this.

That's a good idea. Does it mean we would check a boolean and `compact_pages_rewritten_expected`?

Reply:
> The only question in my mind is whether we could get a call to __wt_block_compact_progress after setting session->compact_state = WT_COMPACT_EXITING but before actually exiting the compaction. If there is such possibility, it would unnecessarily redo the estimation.

After checking the code, it is possible. I could prove it with the reproducer too. We need to check for `WT_COMPACT_EXITING` in more places to exit early.

Reply:
Regarding:

> If there is any doubt, feel free to add another field to WT_BLOCK that marks whether the estimation has been done for the file, which would be a more robust way to do this.

I have added a new flag in 108a364498b8542b89aa09ac3c113257547fd0a5, let me know what you think and if you had something different in mind.

Reply:
I think this makes sense; I don't think there is a need to check `compact_pages_rewritten_expected` in addition to the boolean—just change line 544 to check for `compact_estimated` instead.

Reply:
Sure, done in 62b2b6a92d1ad8d2c861116e1d8703395e4ca7ad.

Reply:
I'm wondering if there's a point where we should redo the estimation phase here. If `block->compact_pages_rewritten > block->compact_pages_rewritten_expected`. The estimation phase was probably off and we should redo it. Although we should only check this after we have estimated at least once to ensure `block->compact_pages_rewritten_expected > 0`.

Reply:
I agree, I think if the actual work is not aligning with the estimation, it would be worth triggering the estimation.

Reply:
I have created [WT-12332](https://jira.mongodb.org/browse/WT-12332).

Reply:
Cheers

Reply:
There was already a ticket for it, my bad: WT-12319.